### PR TITLE
Add `ConditionSelectorValue`.

### PR DESCRIPTION
**Pull request by hatyo**
This introduces `ConditionSelectorValue` which can be used in conjunction with `PickValue` to implement features such as SQL `case-when`

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ConditionSelectorValue.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.List;
+
+/**
+ * This returns the index of the first implication that is satisfied.
+ *
+ * @see PickValue for more information.
+ */
+public class ConditionSelectorValue extends AbstractValue {
+
+    private static final ObjectPlanHash BASE_HASH = new ObjectPlanHash("Implication-Selector-Value");
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ConditionSelectorValue.java:42]: Can you make this the same-ish as the class name?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ConditionSelectorValue.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            if (Boolean.TRUE.equals(result)) {
+                return i;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public String toString() {
+        return "ImplicationSelector";
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ConditionSelectorValue.java:89]: same here

**Review by normen662 — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ConditionSelectorValue.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    @Override
+    public int planHash(@Nonnull final PlanHashKind hashKind) {
+        return PlanHashable.objectsPlanHash(hashKind, BASE_HASH, implications);
+    }
+
+    @Override
+    public int hashCodeWithoutChildren() {
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ConditionSelectorValue.java:98]: shouldn't there also be a `equalsWithoutChildren`?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ConditionSelectorValue.java (hunk 1 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ConditionSelectorValue.java:42]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ConditionSelectorValue.java (hunk 2 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ConditionSelectorValue.java:89]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ConditionSelectorValue.java (hunk 3 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/ConditionSelectorValue.java:98]: It is delegating to `AbstractValue` which is doing it correctly:

```java
    default boolean equalsWithoutChildren(@Nonnull final Value other,
                                          @Nonnull final AliasMap equivalenceMap) {
        if (this == other) {
            return true;
        }

        return other.getClass() == getClass();
    }
```

**Review by normen662 — APPROVED** [review]

**Pull request merged**